src/Symfony/Component/Lock/Store/MysqlStore.php

```diff
@@ -0,0 +1,153 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock\Store;
+
+use Symfony\Component\Lock\Exception\InvalidArgumentException;
+use Symfony\Component\Lock\Exception\LockAcquiringException;
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\PersistingStoreInterface;
+
+/**
+ * @author rtek
+ * @author Jérôme TAMARELLE <jerome@tamarelle.net>
+ */
+class MysqlStore implements PersistingStoreInterface
+{
+    private \PDO $conn;
+
+    private string $dsn;
+
+    private array $options;
+
+    private int $connectionId;
+
+    public function __construct(\PDO|string $connOrDsn, array $options = [])
+    {
+        if ($connOrDsn instanceof \PDO) {
+            $this->conn = $connOrDsn;
+            $this->assertMysqlDriver();
+            if (\PDO::ERRMODE_EXCEPTION !== $this->conn->getAttribute(\PDO::ATTR_ERRMODE)) {
+                throw new InvalidArgumentException(sprintf('"%s" requires PDO error mode attribute be set to throw Exceptions (i.e. $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION)).', __METHOD__));
+            }
+        } else {
+            $this->dsn = $connOrDsn;
+        }
+
+        $this->options = $options;
+    }
+
+    public function save(Key $key): void
+    {
+        if ($this->exists($key)) {
+            return;
+        }
+
+        $name = self::getLockName($key);
+        $stmt = $this->getConnection()->prepare('SELECT IF(IS_USED_LOCK(:name) = CONNECTION_ID(), -1, GET_LOCK(:name, 0))');
+        $stmt->bindValue(':name', $name, \PDO::PARAM_STR);
+        $stmt->execute();
+        $result = $stmt->fetchColumn();
+
+        // lock acquired
+        if (1 === $result) {
+            $key->setState($this->getStateKey($key), $name);
+
+            return;
+        }
+
+        if (0 === $result) {
+            throw new LockConflictedException('Lock already acquired by other connection.');
+        }
+
+        if (-1 === $result) {
+            throw new LockConflictedException('Lock already acquired by this connection.');
+        }
+
+        throw new LockAcquiringException('Failed to acquire lock due to mysql error.');
+    }
+
+    public function putOffExpiration(Key $key, float $ttl): void
+    {
+        // noop - GET_LOCK() does not have a ttl
+    }
+
+    public function delete(Key $key): void
+    {
+        $stmt = $this->getConnection()->prepare('DO RELEASE_LOCK(:name)');
+        $stmt->bindValue(':name', self::getLockName($key), \PDO::PARAM_STR);
+        $stmt->execute();
+
+        $key->removeState($this->getStateKey($key));
+    }
+
+    public function exists(Key $key): bool
+    {
+        $stateKey = $this->getStateKey($key);
+        if (!$key->hasState($stateKey)) {
+            return false;
+        }
+
+        $stmt = $this->getConnection()->prepare('SELECT IF(IS_USED_LOCK(:name) = CONNECTION_ID(), 1, 0)');
```

Comment:
this prepares the statement each time the method is called, consider moving it to after the connection is established and capturing statement in an object property

Reply:
IF statement is technically not needed, return the result of comparison
```suggestion
        $stmt = $this->getConnection()->prepare('SELECT IS_USED_LOCK(:name) = CONNECTION_ID()');
```

you would get the `1` when the lock is held, and `null` when it's not (safer to assume not `1` value)

Reply:
agree